Update date time ref

BillWagner opened the pull request:
Fixes #3502 
[Internal review link](https://review.docs.microsoft.com/en-us/dotnet/api/system.datetime?branch=pr-en-us-3799)

BillWagner committed:
move examples section up

It appears before remarks anyway, as part of the build process. Put it that way in the file to make editing easier.

BillWagner committed:
tab all sample code

Time for test build

BillWagner committed:
fix buid errors

JRAlexander commented on xml/System/DateTime.xml:
```diff
@@ -79,54 +89,80 @@
-  
+    # [Visual Basic](#tab/visual-basic)
+    [!code-vb[System.DateTime.Instantiation#2](~/samples/snippets/visualbasic/VS_Snippets_CLR_System/system.DateTime.Instantiation/vb/Instantiation1.vb#2)]  
+    ---
```
Should there be a C# equivalent for this?

BillWagner replied:
Nope, this is only a VB thing.

JRAlexander reviewed (dismissed):
One small nit, otherwise LGTM.

mairaw commented on xml/System/DateTime.xml:
```diff
@@ -46,7 +46,17 @@
+## Examples  
+ The following example demonstrates how to compare roughly equivalent <xref:System.DateTime> values, accepting a small margin of difference when declaring them equal.  
   
+# [C++](#tab/cpp)
```
why are we using tabs for the sample code? they can already select the language on the language selector, no?

BillWagner committed:
test build after updating samples

BillWagner committed:
Create a single downloadable sample.

BillWagner committed:
fix build errors

BillWagner committed:
Add the last code sample

BillWagner committed:
review and edit

BillWagner commented:
@JRAlexander @mairaw @rpetrusha This is ready for review

svick commented on xml/System/DateTime.xml:
```diff
@@ -46,351 +46,353 @@
+
+### DateTime values and their string representations  
+
+Internally, all <xref:System.DateTime> values are represented as the number of ticks (the number of 100-nanosecond intervals) that have elapsed since 12:00:00 midnight, January 1, 0001. The actual <xref:System.DateTime> value is independent of the way in which that value appears when displayed. The appearance of a <xref:System.DateTime> value is the string result of a formatting operation that co ...
```
Extra "string" in "is the string result"?

svick commented on xml/System/DateTime.xml:
```diff
@@ -46,351 +46,353 @@
+Internally, all <xref:System.DateTime> values are represented as the number of ticks (the number of 100-nanosecond intervals) that have elapsed since 12:00:00 midnight, January 1, 0001. The actual <xref:System.DateTime> value is independent of the way in which that value appears when displayed. The appearance of a <xref:System.DateTime> value is the string result of a formatting operation that co ...
   
- UTC time is suitable for calculations, comparisons, and storing dates and time in files. Local time is appropriate for display in user interfaces of desktop applications. Time zone-aware applications (such as many Web applications) also need to work with a number of other time zones.  
+The appearance of date and time values is dependent on culture, international standards, application requirements, and personal preference. The <xref:System.DateTime> structure offers flexibility in formatting date and time values through overloads of  <xref:System.DateTime.ToString%2A>. The default <xref:System.DateTime.ToString?displayProperty=nameWithType> method returns the string representat ...
```
Shouldn't cultures be written with the country in capitals? E.g. "en-US" here, "fr-FR" below?

BillWagner replied:
That's what I originally thought. Paging @mairaw  Acrolinx mandated all lower case.

rpetrusha replied:
Although they're actually case-insensitive when supplied to the CultureInfo constructors and methods, country should be uppercase, @BIllWagner; the strings are defined by an international standard (not that I remember, of course, what the standard document is).

svick commented on xml/System/DateTime.xml:
```diff
@@ -46,351 +46,353 @@
   
 1.  If the values represent single moments of time, convert them from the local time to UTC by calling the <xref:System.DateTime.ToUniversalTime%2A> method.  
+1.  Convert the dates to their string representations by calling the <xref:System.DateTime.ToString%28System.String%2CSystem.IFormatProvider%29> or <xref:System.String.Format%28System.IFormatProvider%2CSystem.String%2CSystem.Object%5B%5D%29?displayProperty=nameWithType> overload. Use the formatting conventions of the invariant culture by specifying <xref:System.Globalization.CultureInfo.Invariant ...
+1.  When you call the <xref:System.DateTime.Parse%28System.String%2CSystem.IFormatProvider%2CSystem.Globalization.DateTimeStyles%29> or <xref:System.DateTime.TryParse%28System.String%2CSystem.IFormatProvider%2CSystem.Globalization.DateTimeStyles%2CSystem.DateTime%40%29> method.  
```
This sentence seems to be incomplete: "When you call the `DateTime.Parse(string, IFormatProvider, DateTimeStyles)` or `DateTime.TryParse(string, IFormatProvider, DateTimeStyles, out DateTime)` method."

svick commented on xml/System/DateTime.xml:
```diff
@@ -46,351 +46,353 @@
+
+#### Serializing DateTime values  
+
+You can persist <xref:System.DateTime> values through serialization to a stream or file, and then restore them through deserialization. <xref:System.DateTime> data is serialized in some specified object format. The objects are restored when they are deserialized. A formatter or serializer, such as <xref:System.Xml.Serialization.XmlSerializer> or <xref:System.Runtime.Serialization.Formatters.Binar ...
```
Should this link be updated to `~/standard/serialization/index.md`?

svick commented on xml/System/DateTime.xml:
```diff
@@ -3434,7 +3436,7 @@ juillet 2009
  The value of this constant is equivalent to 23:59:59.9999999 UTC, December 31, 9999 in the Gregorian calendar, exactly one 100-nanosecond tick before 00:00:00 UTC, January 1, 10000.  
   
- Some calendars, such as the <xref:System.Globalization.UmAlQuraCalendar>, support an upper date range that is earlier than <xref:System.DateTime.MaxValue>. In these cases, trying to access <xref:System.DateTime.MaxValue> in variable assignments or formatting and parsing operations can throw an <xref:System.ArgumentOutOfRangeException>. Rather than retrieving the value of <xref:System.DateTime.Ma ...
+ Some calendars, such as the <xref:System.Globalization.ThaiBuddhistCalendar>, support an upper date range that is earlier than <xref:System.DateTime.MaxValue>. In these cases, trying to access <xref:System.DateTime.MaxValue> in variable assignments or formatting and parsing operations can throw an <xref:System.ArgumentOutOfRangeException>. Rather than retrieving the value of <xref:System.DateTim ...
```
What is `System.Globalization.CultureInfo.DateTimeFormat.Calendar.MaxSupportedDateTime` supposed to mean? It's not referring just to the property, but it's not a valid expression either. I think this should just use an xref: `<xref:System.Globalization.Calendar.MaxSupportedDateTime>`.

svick commented on xml/System/DateTime.xml:
```diff
@@ -46,351 +46,353 @@
+
+You can persist <xref:System.DateTime> values through serialization to a stream or file, and then restore them through deserialization. <xref:System.DateTime> data is serialized in some specified object format. The objects are restored when they are deserialized. A formatter or serializer, such as <xref:System.Xml.Serialization.XmlSerializer> or <xref:System.Runtime.Serialization.Formatters.Binar ...
+  
+The following example uses the <xref:System.Xml.Serialization.XmlSerializer> class to serialize and deserialize <xref:System.DateTime> values. The values represent all leap year days in the twenty-first century. The output represents the result if the example is run on a system whose current culture is English (G
```
Why is this information useful?

BillWagner replied:
@rpetrusha I had the same thought as I was updating this. Do you remember? Should it be deleted?

rpetrusha replied:
@BillWagner and @svick, I'm sorry, I don't quite follow. What about line 381, what's showing incorrectly, and what should be deleted? (I'm probably missing something glaringly obvious.)

svick replied:
@rpetrusha Line 381 is:

> ### Version considerations  
> 
> Starting with the .NET Framework 2.0, the `System.DateTime` structure contains a 64-bit field composed of a private Kind field concatenated with the Ticks field. The Kind field is a 2-bit field that indicates whether the `System.DateTime` structure represents a local time, a Coordinated Universal Time (UTC), or the time in an unspecified time zone. The Kind field is used when performing time conversions between time zones, but not for time comparisons or arithmetic. The value of the Kind field can be obtained with the `DateTime.Kind` property.

I don't see why would this information be useful now.

I commented about this on line 381, but for some reason, my comment is incorrectly showing here.

rpetrusha replied:
Thanks for the clarification, @svick. I agree, @BillWagner, that it should be removed.

svick commented on xml/System/DateTime.xml:
```diff
@@ -46,351 +46,353 @@
+
+You can persist <xref:System.DateTime> values through serialization to a stream or file, and then restore them through deserialization. <xref:System.DateTime> data is serialized in some specified object format. The objects are restored when they are deserialized. A formatter or serializer, such as <xref:System.Xml.Serialization.XmlSerializer> or <xref:System.Runtime.Serialization.Formatters.Binar ...
+  
+The following example uses the <xref:System.Xml.Serialization.XmlSerializer> class to serialize and deserialize <xref:System.DateTime> values. The values represent all leap year days in the twenty-first century. The output represents the result if the example is run on a system whose current culture is English (G
```
Why is this note part of Version considerations? Shouldn't it be in one of the above sections, like DateTime values or DateTime operations?

BillWagner commented on xml/System/DateTime.xml:
```diff
@@ -46,351 +46,353 @@
+ [Persisting DateTime values](#persisting-datetime-values)   
+ [Comparing for equality within tolerance](#comparing-for-equality-within-tolerance)
+
+You can download the entire set of examples from this article in either [Visual Basic](https://github.com/docs/samples/snippets/visualbasic/System.DateTime/) or [C#](https://github.com/docs/samples/snippets/csharp/System.DateTime/) from the docs repository on GitHub.
```
We are investigating DownGit now.  What do you think of the experience? There are other alternatives we are looking at as well. /cc @dend

svick replied:
Not sure if it's DownGit's fault, but when I download the zip using DownGit, my Firefox tells me that "This file is not commonly downloaded" and that it may not be safe to open:

![](https://user-images.githubusercontent.com/287848/33572436-38439896-d933-11e7-961e-0bd2175e894b.png)

If there is another service where this doesn't happen, I would prefer that.

BillWagner committed:
respond to feedback

BillWagner committed:
respond to feedback

Fix links to sample code.
Remove description of internal storage.

rpetrusha commented on xml/System/DateTime.xml:
```diff
@@ -46,351 +46,351 @@
   
- [!code-csharp[System.DateTime.Persist#1](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.datetime.persist/cs/persist1.cs#1)]
- [!code-vb[System.DateTime.Persist#1](~/samples/snippets/visualbasic/VS_Snippets_CLR_System/system.datetime.persist/vb/persist1.vb#1)]  
+ The most common error made when persisting <xref:System.DateTime> values as strings is to rely on the formatting conventions of the default or current culture. Problems arise if the current culture is different when saving and restoring the strings. The following example illustrates these problems. It saves five dates using the formatting conventions of the current culture, which in this case is ...
```
formatting conventions a --> formatting conventions of a

rpetrusha commented on xml/System/DateTime.xml:
```diff
@@ -46,351 +46,351 @@
+To round-trip <xref:System.DateTime> values successfully, follow these steps:  
   
 1.  If the values represent single moments of time, convert them from the local time to UTC by calling the <xref:System.DateTime.ToUniversalTime%2A> method.  
+1.  Convert the dates to their string representations by calling the <xref:System.DateTime.ToString%28System.String%2CSystem.IFormatProvider%29> or <xref:System.String.Format%28System.IFormatProvider%2CSystem.String%2CSystem.Object%5B%5D%29?displayProperty=nameWithType> overload. Use the formatting conventions of the invariant culture by specifying <xref:System.Globalization.CultureInfo.Invariant ...
```
the "O" or "R". --> the "O" or "R" standard format string.

rpetrusha commented on xml/System/DateTime.xml:
```diff
@@ -46,351 +46,351 @@
   
 1.  If the values represent single moments of time, convert them from the local time to UTC by calling the <xref:System.DateTime.ToUniversalTime%2A> method.  
+1.  Convert the dates to their string representations by calling the <xref:System.DateTime.ToString%28System.String%2CSystem.IFormatProvider%29> or <xref:System.String.Format%28System.IFormatProvider%2CSystem.String%2CSystem.Object%5B%5D%29?displayProperty=nameWithType> overload. Use the formatting conventions of the invariant culture by specifying <xref:System.Globalization.CultureInfo.Invariant ...
+1.  Call the <xref:System.DateTime.Parse%28System.String%2CSystem.IFormatProvider%2CSystem.Globalization.DateTimeStyles%29> or <xref:System.DateTime.TryParse%28System.String%2CSystem.IFormatProvider%2CSystem.Globalization.DateTimeStyles%2CSystem.DateTime%40%29> method.  
```
I think that this can be deleted. This relates to the parsing operation, which is described in the next paragraph and numbered list.

rpetrusha commented on xml/System/DateTime.xml:
```diff
@@ -46,351 +46,351 @@
+
+#### Persisting values as integers  
+
+You can persist it as an <xref:System.Int64> value that represents a number of ticks. In this case, you don't have to consider the culture of the systems the <xref:System.DateTime> values are persisted and restored on.  
```
it is an undefined pronoun here. persist it --> persist a date and time

rpetrusha commented on xml/System/DateTime.xml:
```diff
@@ -46,351 +46,351 @@
+
+You can persist <xref:System.DateTime> values through serialization to a stream or file, and then restore them through deserialization. <xref:System.DateTime> data is serialized in some specified object format. The objects are restored when they are deserialized. A formatter or serializer, such as <xref:System.Xml.Serialization.XmlSerializer> or <xref:System.Runtime.Serialization.Formatters.Binar ...
+  
+The following example uses the <xref:System.Xml.Serialization.XmlSerializer> class to serialize and deserialize <xref:System.DateTime> values. The values rep
```
1. deserialize it convert --> deserialize it, convert
2.  I think it would be more useful for the example to use something other than UTC, so that it retrieves a UTC time and then has to call ToLocalTime and then displays it in a different time zone . For example, here using US Central Standard Time should produce a difference of 2 hours from the original time.

rpetrusha replied:
There's a lack of context here. Mention that equality comparisons are exact -- two times are equal if they are expressed as the same number of ticks.

rpetrusha replied:
This isn't really a version consideration, it's an alternative. I think I'd remove the section, and move the box up to immediately precede the first subhead in the Examples section.

rpetrusha replied:
The Remarks section (the head) is not rendering. the mini TOC that follows is. But effectively, all of this content is not discoverable.
I'm not sure why, but one possible explanation is that in the markdown, the parser expects Remarks to precede Examples, since that's the way the DDUEML was formatted, and possibly how EXMAXML expects mref to be formatted. 
//cc @dend

rpetrusha replied:
no comma after "represents the date"

rpetrusha replied:
Some offsets are not even hours. So change "adds or subtracts an hour" to "adds or subtracts a time interval adjustment". "Adds or subtracts an hour from the length of a day" makes it sound as if a day can be from 23 to 25 hours, which is true only of the days on which an adjustment goes into effect.

rpetrusha replied:
positive, can --> positive, and they can

rpetrusha replied:
comma before while

rpetrusha replied:
1. deserialize it convert --> deserialize it, convert
2. I think it would be more useful for the example to use something other than UTC, so that it retrieves a UTC time and then has to call ToLocalTime and then displays it in a different time zone . For example, here using US Central Standard Time should produce a difference of 2 hours from the original time.

rpetrusha commented on xml/System/DateTime.xml:
```diff
@@ -46,351 +46,351 @@
+
+You can persist <xref:System.DateTime> values through serialization to a stream or file, and then restore them through deserialization. <xref:System.DateTime> data is serialized in some specified object format. The objects are restored when they are deserialized. A formatter or serializer, such as <xref:System.Xml.Serialization.XmlSerializer> or <xref:System.Runtime.Serialization.Formatters.Binar ...
+  
+The following example uses the <xref:System.Xml.Serialization.XmlSerializer> class to serialize and deserialize <xref:System.DateTime> values. The values rep
```
This can be deleted.

BillWagner replied:
What do you think can be deleted? (You don't mean the whole paragraph, do you?)

rpetrusha reviewed (changes requested):
Left a number of comments, @BillWagner. Also, I'm wondering if Examples and Remarks can be integrated in some way.

BillWagner commented on xml/System/DateTime.xml:
```diff
@@ -46,351 +46,351 @@
     <summary>Represents an instant in time, typically expressed as a date and time of day.</summary>
     <remarks>
       <format type="text/markdown"><![CDATA[  
+## Examples
```
I agree @rpetrusha  I didn't like the formatting either.  

What do you think of removing "examples" as a header, and having sub-headings under "remarks" for each scenario? In each sub-heading, explain the scenario and have the examples.  

If I remove "Examples" as an H2, that should prevent the build process from re-ordering, right?

rpetrusha replied:
@BillWagner, it should prevent the reordering, but I think that we may just have to play around with levels of heads and see how the build turns out. What we also have to do is make sure that we're not taking a dependency on a build feature that is likely to change. (@dend can help here). 
But I think it's worthwhile removing remarks, and then maybe promoting all level-3 heads to level-2 heads, and adding at the end an examples-like section (perhaps `## The example code`  or something similar) that mentions that the examples are interspersed throughout, and can be downloaded from the link.
Given the experience with many of the mref topics where there is no examples section but dozens of examples in the Remarks, where the most common customer complaint was that there were no examples, we have to do something to try to make clear that the documentation has both plenty of examples and plenty of remarks.

BillWagner commented on xml/System/DateTime.xml:
```diff
@@ -46,351 +46,351 @@
+ [Persisting DateTime values](#persisting-datetime-values)   
+ [Comparing for equality within tolerance](#comparing-for-equality-within-tolerance)
+
+You can download the entire set of examples from this article in either [Visual Basic](https://github.com/dotnet/docs/tree/master/samples/snippets/visualbasic/System.DateTime/) or [C#](https://github.com/dotnet/docs/tree/master/samples/snippets/csharp/System.DateTime/) from the docs repository on GitHub.
```
I changed the wording to "view the source for" from "download"

BillWagner commented on xml/System/DateTime.xml:
```diff
@@ -46,351 +46,351 @@
+[!code-csharp-interactive[System.DateTime.Formatting#3](~/samples/snippets/csharp/System.DateTime/StringFormat.cs#3)]
+[!code-vb[System.DateTime.Formatting#3](~/samples/snippets/visualbasic/System.DateTime/StringFormat.vb#3)]  
+
+ Finally, you can specify both the culture and the representation using the <xref:System.DateTime.ToString%28System.String%2CSystem.IFormatProvider%29?displayProperty=nameWithType> method. The following example uses the <xref:System.DateTime.ToString%28System.String%2CSystem.IFormatProvider%29?displayProperty=nameWithType> method to display the full date and time pattern for the fr-FR culture.  
```
I felt "format" was better.

BillWagner commented on xml/System/DateTime.xml:
```diff
@@ -46,351 +46,351 @@
+[!code-csharp-interactive[System.DateTime.Parsing#2](~/samples/snippets/csharp/System.DateTime/Parsing.cs#2)]
+[!code-vb[System.DateTime.Parsing#2](~/samples/snippets/visualbasic/System.DateTime/Parsing.vb#2)]  
+
+One common use for the <xref:System.DateTime.Parse%2A> and <xref:System.DateTime.ParseExact%2A> is to convert a string representation from a web service, usually in [ISO 8601](https://www.iso.org/iso-8601-date-and-time-format.html) standard format. The following code shows the correct format string to use:
```
Good catch.

rpetrusha commented on xml/System/DateTime.xml:
```diff
@@ -46,351 +46,351 @@
+
+You can persist <xref:System.DateTime> values through serialization to a stream or file, and then restore them through deserialization. <xref:System.DateTime> data is serialized in some specified object format. The objects are restored when they are deserialized. A formatter or serializer, such as <xref:System.Xml.Serialization.XmlSerializer> or <xref:System.Runtime.Serialization.Formatters.Binar ...
+  
+The following example uses the <xref:System.Xml.Serialization.XmlSerializer> class to serialize and deserialize <xref:System.DateTime> values. The values rep
```
The Remarks section (the head) is not rendering. the mini TOC that follows is. But effectively, all of this content is not discoverable.
I'm not sure why, but one possible explanation is that in the markdown, the parser expects Remarks to precede Examples, since that's the way the DDUEML was formatted, and possibly how EXMAXML expects mref to be formatted.
//cc @dend

mairaw replied:
I think that Remarks need to come first in the file and they will be rendered after.

rpetrusha commented:
@BillWagner, I think that I've added line numbers to all the comments that appeared on an incorrect line. Let me know if there are any that I missed.

BillWagner committed:
respond to feedback

This commit responds to the editing feedback.
The next commit will do the re-arranging for remarks / examples and how to present the information

BillWagner committed:
test: reorder

Remove the "Remarks" and "Examples" headers.
Promote other headers 1 level.

BillWagner commented:
@rpetrusha @mairaw I updated and re-ordered content based on the above discussion. It's ready for another look.

rpetrusha commented:
In terms of content, this looks really good, @BillWagner. In terms of structure, it's problematic. What's happened is that, despite the fact that you've removed the \#\# Remarks  head, the build has supplied one, and has evidently demoted your level-2 heads to level-3. There's also no examples section, so many developers who just look for the Examples section for examples are going to miss the innumerable examples here. I think that structurally we probably need to base the structure more on your earlier PR, and try to tweak it a little bit... It somehow seems clear, though, that actually having an examples section and placing it before remarks disorients the build and causes it to be more literal in its treatment of sections and subsections.

BillWagner committed:
add example section with links to anchors

This provides easy navigation to all code samples.

BillWagner commented:
@rpetrusha @mairaw I updated based on our discussion Friday. I think this looks good. Let me know what you think

BillWagner committed:
format links to examples.

Add the space before each line to get each link on a different line.

rpetrusha commented:
@BillWagner, a number of comments:
- There's one broken link: `Invalid file link:(~/standard/serialization/index.md)` (should be ~/docs/standard...).
- The link to examples in the Examples section all appear on the same line. Each link should be terminated by two spaces.
- This didn't produce the mini-TOC that we were hoping for, though we at least do now have Remarks and Examples, and detailed links in the examples.
Rather than a new round of adjustments, which may not accomplish anything, I think that this PR is ready to merge once the line formatting and the link is fixed. So I'll approve the PR.

BillWagner committed:
fix broken link.

BillWagner committed:
fix example formatting.

BillWagner committed:
trying that formatting update again

And, headers named "Remarks" and "examples" are removed at build time. Put different headers in place.

BillWagner committed:
fix anchor tags
